tanjun/parsing.py

```diff
@@ -31,8 +31,321 @@
+        self.__shlex.whitespace = " "
+        self.__shlex.whitespace_split = True
+
+    def collect_options(self) -> typing.Mapping[str, typing.Sequence[typing.Optional[str]]]:
+        results: typing.MutableMapping[str, typing.MutableSequence[typing.Optional[str]]] = {}
+
+        while (option_ := self.next_option()) is not None:
+            name, value = option_
+
+            if name not in results:
+                results[name] = []
+
+            results[name].append(value)
+
+        return results
+
+    def iter_arguments(self) -> typing.Iterator[str]:
+        while (argument_ := self.next_argument()) is not None:
+            yield argument_
+
+    def next_argument(self) -> typing.Optional[str]:
+        if self.__arg_buffer:
+            return self.__arg_buffer.pop(0)
+
+        while isinstance(value := self.__seek_shlex(), tuple):
+            self.__options_buffer.append(value)
+
+        return value
+
+    def next_option(self) -> typing.Optional[typing.Tuple[str, typing.Optional[str]]]:
+        if self.__options_buffer:
+            return self.__options_buffer.pop(0)
+
+        while isinstance(value := self.__seek_shlex(), str):
+            self.__arg_buffer.append(value)
+
+        return value
+
+    def __seek_shlex(self) -> typing.Union[str, typing.Tuple[str, typing.Optional[str]], None]:
+        option_name = self.__last_name
+        for value in self.__shlex:
+            is_option = value.startswith("-")
+            if is_option and option_name is not None:
+                self.__last_name = value
+                return (option_name, None)
+
+            if is_option:
+                option_name = value
+                continue
+
+            if option_name:
+                return (option_name, value)
+
+            return value
+
+        if self.__last_name is not None:
+            last_name = self.__last_name
+            self.__last_name = None
+            return (last_name, None)
+
+        return None
+
+
+async def _convert_or_default_option(
+    ctx: traits.Context, option_: traits.Parameter, value: typing.Optional[typing.Any], /
+) -> typing.Any:
+    if value is not None:
+        return await option_.convert(ctx, value)
+
+    if option_.empty_value is not undefined.UNDEFINED:
+        return option_.empty_value
+
+    raise errors.ParserError(f"Option '{option_.names[0]} cannot be empty.", option_)
+
+
+class SemanticShlex(ShlexTokenizer):
+    __slots__: typing.Sequence[str] = ("__ctx",)
+
+    def __init__(self, ctx: traits.Context, /) -> None:
+        super().__init__(ctx.content)
+        self.__ctx = ctx
+
+    async def get_arguments(self, arguments: typing.Sequence[traits.Parameter], /) -> typing.Sequence[typing.Any]:
+        results: typing.MutableSequence[typing.Any] = []
+        for argument_ in arguments:
+            results.append(await self.__process_argument(argument_))
+
+            if argument_.flags.get("greedy") or argument_.flags.get("multi"):
```

Comment:
Might want to make GREEDY and MULTI documented constants and maybe also add decorators for them specifically